### PR TITLE
fuzzgen: Add scalar float support

### PR DESCRIPTION
**Pull request by afonso360**
👋  Hey,

This PR adds support for generating floats in fuzzgen and adds a bunch of instructions.

I think we have all the float ops, except bit ops that operate on floats.

So far this has found #4446 and #4460. We can merge this before #4460 but it will probably start crashing in oss-fuzz.

cc: @jameysharp

### CODE REVIEW

**Review by jameysharp — CHANGES_REQUESTED** [review]
I think this looks great and I'm excited to get more fuzzing for Cranelift!

I saw one copy-paste bug and sort of expect I might have missed some, because having a separate function for each arity leaves a lot of opportunities for this kind of bug. I don't mind merging this with just the one bug fixed, but if you're willing to take a little time to see if you can refactor that, that would be cool.

I'm trying to decide whether to wait on #4460 before merging this. I don't know how long that will take to get reviewed. If you're currently only seeing bugs in the `Fma` opcode, how would you feel about commenting out that opcode in this patch? Then there's no question, I'm happy to merge it immediately.

File: cranelift/fuzzgen/src/function_generator.rs [hunk 1]
```diff
@@ -49,6 +70,33 @@ fn insert_opcode_arity_2(
...
+    args: &'static [Type],
+    rets: &'static [Type],
+) -> Result<()> {
+    let arg0 = fgen.get_variable_of_type(args[0])?;
+    let arg0 = builder.use_var(arg0);
+
+    let arg1 = fgen.get_variable_of_type(args[1])?;
+    let arg1 = builder.use_var(arg1);
+
+    let arg2 = fgen.get_variable_of_type(args[1])?;
```
[the remark this answers is not in the document]
Comment (afonso360) [reply cranelift/fuzzgen/src/function_generator.rs:86]: Good Catch! I missed the existence of `MultiAry` but it looks like a much better solution. I'll replace it.

**Comment by afonso360** [comment]
>  If you're currently only seeing bugs in the Fma opcode, how would you feel about commenting out that opcode in this patch? Then there's no question, I'm happy to merge it immediately.

Sure, sounds good. I've commented it out and left a link on it.

**Review by jameysharp — APPROVED** [review]
This looks great, thank you!

File: cranelift/fuzzgen/src/function_generator.rs [hunk 2]
```diff
@@ -10,40 +11,32 @@ use cranelift::frontend::{FunctionBuilder, FunctionBuilderContext, Variable};
...
     opcode: Opcode,
     args: &'static [Type],
     rets: &'static [Type],
 ) -> Result<()> {
-    let arg0 = fgen.get_variable_of_type(args[0])?;
-    let arg0 = builder.use_var(arg0);
-
-    let arg1 = fgen.get_variable_of_type(args[1])?;
-    let arg1 = builder.use_var(arg1);
+    let mut pool = ListPool::new();
```
Comment (cfallin) [inline cranelift/fuzzgen/src/function_generator.rs:26]: Hmm, I'm surprised that this works, and I suspect it does only accidentally: we should be building the argument list in the pool owned by the `DataFlowGraph`. The `ValueList` is just an index into the pool, so `ValueList`s need to be built with the proper pool and will silently fail if referring to another pool. (This is a kind of unfortunate logical-unsafety but the type system doesn't give us a way to tie the list to a *particular* pool.)

Comment (afonso360) [reply cranelift/fuzzgen/src/function_generator.rs]: Fixed. I wonder how the fuzzer didn't immediately catch this, I only ran it for a couple of minutes after making this change, but I would have expected it to fail instantly.

Thanks!

File: cranelift/fuzzgen/Cargo.toml [hunk 3]
```diff
@@ -12,6 +12,8 @@ publish = false
 
 [dependencies]
 cranelift = { path = "../umbrella", version = "0.87.0" }
+cranelift-native = { path = "../native", version = "0.87.0" }
+cranelift-entity = { path = "../entity", version = "0.87.0" }
```
Comment (jameysharp) [inline cranelift/fuzzgen/Cargo.toml:16]: Thanks for fixing the use of `ValueList`! I think you don't need a dependency on `cranelift-entity` now, right?

Comment (afonso360) [reply cranelift/fuzzgen/Cargo.toml]: Right, Fixed!

**Comment by jameysharp** [comment]
Fantastic! We're hitting CI hard today so this probably won't get a chance for the CI checks to run for a while, but I'm setting it to auto-merge once that passes. Thanks for your efforts on this!

I think I understand why fuzzing didn't catch the misuse of `ListPool`: since you created a new pool, all the values in your `ValueList`s had small indexes (0..3) into the pool, and it's not so surprising that those were valid indexes into the DFG's pool as well. Or something like that.

**Comment by alexcrichton** [comment]
Apologies for the CI breakage, but things should be fixed with a rebase.

**Pull request merged**